File: Data Structures & Algorithms/capacity-to-ship-packages-within-d-days/submission-0.cpp
class Solution {
public:
    int shipWithinDays(vector<int>& weights, int days) {
        int lo=0;
        int hi=0;
        for(int i=0;i<weights.size();i++){
            hi+=weights[i];
            if(lo<weights[i]){
                lo=weights[i];
            }
        }int mid;int ans=-1;
        while(lo<=hi){
            mid=lo+(hi-lo)/2;
            if(check(mid,weights,days)){
                ans=mid;
                hi=mid-1;
            }else{
                lo=mid+1;
            }
        }return ans;
    }
    bool check(int mid,vector<int>& weights,int days){
        int needed=1;int total=0;
        int n=weights.size();
        for(int i=0;i<n;i++){
            if(total+weights[i]<=mid){
                total+=weights[i];
            }else{
                needed++;
                total=weights[i];
            }
        }
        if(needed<=days) return true;
        return false;
    }
};

File: Data Structures & Algorithms/sqrtx/submission-0.java
class Solution {
    public int mySqrt(int x) {
        int lo=0;
        int hi=x;int ans=0;
        while(lo<=hi){
            long mid=lo+(hi-lo)/2;
            long val=mid*mid;
            if(val<=x){
                ans=(int)mid;
                lo=(int)mid+1;
            }else{
                hi=(int)mid-1;
            }
        }
        return ans;
    }
}